### PR TITLE
Add Dictionary Encoded vectors to PyVelox

### PR DESCRIPTION
**Pull request by save-buffer**
This PR adds support for dictionary vectors in PyVelox. Currently base vectors are supported only for primitive (physical) types , and this also thus extends to Dictionary vectors. 
Dictionary vectors are required for additional fuzzer support, and expression eval in pyvelox.

### CODE REVIEW

**Comment by save-buffer** [comment]
@kgpai

File: pyvelox/pyvelox.h [hunk 1]
```diff
@@ -419,9 +516,58 @@ inline void addVectorBindings(
...
+      });
+
   m.def("from_list", [](const py::list& list) mutable {
     return pyListToVector(list, PyVeloxContext::getInstance().pool());
   });
+  m.def("constant_vector", [](const py::handle& obj, vector_size_t length) {
+    return pyToConstantVector(
+        obj, length, PyVeloxContext::getInstance().pool());
+  });
+  m.def("dictionary_vector", [](VectorPtr values, py::list& indices_list) {
```
Comment (kgpai) [inline pyvelox/pyvelox.h:551]: We also need to add null support to the dictionary . Need a utility that also takes nulls buffer.

File: pyvelox/pyvelox.h [hunk 2]
```diff
@@ -368,7 +414,68 @@ inline void addDataTypeBindings(
   rowType.def("names", &RowType::names, "Return the names of the columns");
 }
 
-inline void addVectorBindings(
+struct DictionaryIndices {
```
[the remark this answers is not in the document]
Comment (save-buffer) [reply pyvelox/pyvelox.h:299]: Nope, type alias would still just make object const.

File: pyvelox/pyvelox.h [hunk 3]
```diff
@@ -419,9 +516,58 @@ inline void addVectorBindings(
...
+  });
+  m.def("dictionary_vector", [](VectorPtr values, py::list& indices_list) {
+    BufferPtr indices_buffer = AlignedBuffer::allocate<vector_size_t>(
+        indices_list.size(), PyVeloxContext::getInstance().pool());
+    vector_size_t* indices_ptr = indices_buffer->asMutable<vector_size_t>();
+    for (size_t i = 0; i < indices_list.size(); i++) {
+      if (!py::isinstance<py::int_>(indices_list[i]))
+        throw py::type_error("Found an index that's not an integer");
+      vector_size_t idx = py::cast<vector_size_t>(indices_list[i]);
+      if (idx < 0 || idx >= values->size()) {
```
[the remark this answers is not in the document]
Comment (save-buffer) [reply pyvelox/pyvelox.h:444]: Ah yeah I already have a function for it, not sure why I didn't use it

File: pyvelox/pyvelox.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (save-buffer) [reply pyvelox/pyvelox.h:436]: Changed

File: pyvelox/pyvelox.h [hunk 4]
```diff
@@ -141,6 +182,21 @@ static inline VectorPtr variantsToFlatVector(
   return result;
 }
 
+template <TypeKind T>
+static VectorPtr createDictionaryVector(
+    BufferPtr indices,
+    VectorPtr values,
```
[the remark this answers is not in the document]
Comment (save-buffer) [reply pyvelox/pyvelox.h:126]: Changed

File: pyvelox/pyvelox.h [hunk 5]
```diff
@@ -419,9 +516,58 @@ inline void addVectorBindings(
...
+  py::class_<DictionaryIndices>(
+      m, "DictionaryIndices", py::module_local(asModuleLocalDefinitions))
+      .def(
+          "__len__",
+          [](DictionaryIndices indices) {
+            return (indices.indices->size()) / sizeof(vector_size_t);
+          })
+      .def("__getitem__", [](DictionaryIndices indices, vector_size_t idx) {
+        checkBounds(indices, idx);
+        return indices.indices->as<vector_size_t>()[idx];
```
[the remark this answers is not in the document]
Comment (save-buffer) [reply pyvelox/pyvelox.h:426]: Added `resize` to BaseVector, so now every type of vector can use it

**Review by kgpai — APPROVED** [review]

**Pull request closed without merging**